posts/zzz_DO_NOT_EDIT_the__tensio.../the__tension_of__high-_performance__computing:__reproducibility_vs.__parallelization.qmd

```diff
@@ -0,0 +1,259 @@
+image: "pharmaverse.png"
+
+---
+
+<!--------------- typical setup ----------------->
+
+```{r setup, include=FALSE}
+long_slug <- "zzz_DO_NOT_EDIT_the__tensio..."
+# renv::use(lockfile = "renv.lock")
+```
+
+<!--------------- post begins here ----------------->
+
+## Harnessing HPC for Drug Development
+
+In pharmaceutical research, high-performance computing (HPC) plays a pivotal 
+role in driving advancements in drug discovery and development. From analyzing
+vast genomic datasets to simulating drug interactions across diverse
+populations, HPC enables researchers to tackle complex computational tasks at
+high speeds. As pharmaceutical research becomes increasingly data-driven, the
+need for powerful computational tools has grown, allowing for more accurate
+predictions, faster testing, and more efficient processes. However, with the
+growing complexity and scale of these computations, ensuring reproducibility
+of results becomes a significant challenge.
+
+In this blog post, we will explore common reproducibility challenges in drug
+development and simulations, using the 
+[`{mirai}`](https://shikokuchuo.net/mirai/) package as a backend solution to 
+manage parallelization.
+
+## The Problem: Reproducibility in Parallel Processing
+
+Imagine a research team working on a cutting-edge drug development project. To
+process and analyze vast amounts of data efficiently, they leverage parallel
+processing, distributing tasks across multiple processors. This approach
+significantly accelerates their work, enabling them to handle large datasets and
+complex computations in a fraction of the time.
+
+However, the team soon encounters an issue. Each time they rerun the same
+processing tasks with identical input parameters, the results differ slightly.
+This raises a major concern: *the results are not reproducible.* In industries 
+like pharmaceuticals, where accuracy and consistency are critical,
+reproducibility is not just important—it's a regulatory requirement.
+
+For example, in large-scale Monte Carlo simulations, small differences can arise
+not only from changes in execution order across processors but also from
+inconsistencies between workers or difficulties in maintaining synchronized
+random number generation (RNG) streams. Furthermore, the more complex the
+environment—with multiple components such as distributed workers, different
+hardware, or varying system configurations—the harder it becomes to reprovision
+the exact same environment and repeat the computations exactly. As these
+variations accumulate, ensuring consistent and reproducible results becomes
+a significant challenge in data-driven research.
+
+### Tracking Operations in Parallel Computing
+
+Let’s explore a simple scenario where parallelization creates confusion in 
+tracking operations due to the asynchronous nature of task execution and
+logging. For this, we will also use the 
+[`{tidylog}`](https://github.com/elbersb/tidylog) package, which tracks and logs
```

Comment:
To tie this in with the pharmaverse a bit more, consider using the https://github.com/pharmaverse/logrx package for logging. 
I'd be curious if there is any difference.

Reply:
I agree. 

Used `{tidylog}` + `{logrx}` following this vignette here: https://pharmaverse.github.io/logrx/articles/tidylog.html